xdg-terminal-exec

```diff
@@ -12,295 +12,261 @@
-	IFS=':'
-	debug "lower case desktops are: '${DESKTOPS_LC}'"
-	for CONFIG_DIR in $CONF_HIERARCHY; do
-		for DESKTOP in $DESKTOPS_LC; do
-			printf '%s\n' "${CONFIG_DIR}/${DESKTOP}${DESKTOP:+-}${CONFIG_NAME}"
-		done
-		printf '%s\n' "${CONFIG_DIR}/${CONFIG_NAME}"
-	done | uniq
-	IFS="$OIFS"
-}
-
-find_preferred_entry() {
-	# Read entries listed in config files until a valid one is found
-	CONFIG_LIST="$(generate_config_list)"
-
-	debug 'finding preferred entry in configs'
-
-	while read -r CONFIG_FILE; do
-		debug "looking for config '${CONFIG_FILE}'"
-		if [ -f "${CONFIG_FILE}" ]; then
-			debug "reading config '${CONFIG_FILE}'"
-			while read -r LINE; do
-				debug "parsing line '$LINE'"
-				LINE="$(trim_comment "$LINE")"
-				LINE="$(trim_spaces "$LINE")"
-				[ -z "$LINE" ] && continue || ENTRY_ID="$LINE"
-				debug "retrieved entry ID '$ENTRY_ID'"
-				check_entry_id "$ENTRY_ID" || continue
-
-				debug "finding path for entry ID '$ENTRY_ID'"
-				ENTRY_PATH="$(find_entry_path "$ENTRY_ID")"
-				if [ -n "$ENTRY_PATH" ] && check_entry_path "$ENTRY_PATH"; then
-					debug "file at path '$ENTRY_PATH' checks out"
-					printf '%s' "$ENTRY_PATH"
-					return 0
-				else
-					debug "entry ID '$ENTRY_ID' failed check at path '$ENTRY_PATH', blacklisting"
-					BLACKLIST="${BLACKLIST}${BLACKLIST:+;}$ENTRY_ID"
-					continue
-				fi
-			done < "${CONFIG_FILE}"
-		fi
-	done <<- EOF
-		$CONFIG_LIST
-	EOF
+trim_start() {
+	# Remove all but leading whitespace, and trim that from the given string
+	printf '%s' "${1#"${1%%[![:space:]]*}"}"
 }
 
-find_any_entry() {
-	# Read entries in data dirs until a valid one is found
-	debug 'looking for first available entry in data hierarchy'
-	IFS=':'
-	for DATA_DIR in $DATA_HIERARCHY; do
-		IFS="$OIFS"
-		if [ -d "${DATA_DIR}/${DATA_PREFIX_DIR}" ]; then
-			debug "searching in '${DATA_DIR}/${DATA_PREFIX_DIR}'"
-			ENTRY_FILES="$(find -L "${DATA_DIR}/${DATA_PREFIX_DIR}" -type f -iname '*.desktop')"
-			while read -r ENTRY_PATH; do
-				[ -z "$ENTRY_PATH" ] && continue
-				ENTRY_FILE="${ENTRY_PATH#"${DATA_DIR}/${DATA_PREFIX_DIR}/"}"
-				ENTRY_ID="$(printf '%s' "$ENTRY_FILE" | tr '/' '-')"
-				check_entry_id "$ENTRY_ID" || continue
-				if check_entry_path "$ENTRY_PATH"; then
-					debug "file at path '$ENTRY_PATH' checks out"
-					printf '%s' "$ENTRY_PATH"
-					return 0
-				else
-					debug "entry ID '$ENTRY_ID' failed check at path '$ENTRY_PATH', blacklisting"
-					BLACKLIST="${BLACKLIST}${BLACKLIST:+;}$ENTRY_ID"
-					debug "current blacklist: '$BLACKLIST'"
-				fi
-			done <<- EOF
-				$ENTRY_FILES
-			EOF
-		fi
-	done
-	IFS="$OIFS"
+# Read config from given path, print sanitised entries
+read_config_path() {
+	config_path="$1"
+
+	debug "reading config '$config_path'"
+	[ -f "$config_path" ] || return 0
+	# Let `read` trim leading/trailing whitespace from the line
+	while read -r line; do
+		debug "read line '$line'"
+		case $line in
+		# Comment, skip
+		'#'*) ;;
```

Comment:
this does not drop lines starting with whitespace like `  # comment`

Reply:
You're right, `$IFS` was left at ':', fixed

Reply:
[Desktop entry spec](https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-1.1.html#comments) doesn't mention inline comments, I don't think we should support those

Reply:
`'#'*|'')` for empty lines too

```
	while read -r line; do
		debug "read line '$line'"
		case $line in
		# Comment, skip
		'#'*|'') debug "drop line" ;;
		*) debug "use line" ; printf '%s\n' "$line" ;;
		esac
	done < "$config_path"
```

Reply:
Empty lines should be skipped by the for loop on line 247, do you want to be explicit about it?

Reply:
OK, it's fine, but explicit actions are preferable if not too costly.

Reply:
[According to the spec](https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#file-naming) something like this should work
```
while read -r line; do
	debug "read line '$line'"
	case $line in
	# A desktop entry ID, print it
	[A-Za-z0-9-_]*) printf '%s\n' "$line" ;;
	esac
	# By default empty lines and comments get ignored
done < "$config_path"
```

Reply:
`[A-Za-z_]*) printf '%s\n' "$line" ;;` looks great